File: core/src/Pipeline.ts
import { Plugin, ContentData, PipelineConfig, OutputHandler } from './types';
import { PluginRegistry } from './PluginRegistry';
import { WebRequest, WebResponse, WebOutputAdapter, WebStreamHandler } from './lib/WebAdapter';

export class Pipeline {
  private plugins: Plugin[] = [];
  private outputHandler?: OutputHandler;
  private webAdapter?: WebOutputAdapter;

  constructor(plugins: Plugin[] = [], outputHandler?: OutputHandler) {
    this.plugins = plugins;
    this.outputHandler = outputHandler || this.defaultOutputHandler;
  }

  private defaultOutputHandler: OutputHandler = async (output) => {
    if (this.webAdapter) {
      await this.webAdapter.handleOutput(output);
    } else if (output.type === 'error') {
      console.error(output.content);
    } else {
      console.log(JSON.stringify(output.content, null, 2));
    }
  };

  static async fromConfig(configPath: string, outputHandler?: OutputHandler): Promise<Pipeline> {
    const config = await import(configPath) as PipelineConfig;
    return Pipeline.fromConfigObject(config, outputHandler);
  }

  static fromConfigObject(config: PipelineConfig, outputHandler?: OutputHandler): Pipeline {
    const registry = PluginRegistry.getInstance();
    const plugins = config.plugins.map(pluginConfig => {
      const mergedConfig = {
        ...config.globalOptions,
        ...pluginConfig.options
      };
      const plugin = registry.create(pluginConfig.name, mergedConfig);
      if (plugin.configure && pluginConfig.options) {
        plugin.configure(pluginConfig.options);
      }
      return plugin;
    });
    
    return new Pipeline(plugins, outputHandler);
  }

  addPlugin(plugin: Plugin): Pipeline {
    this.plugins.push(plugin);
    return this;
  }

  setOutputHandler(handler: OutputHandler): Pipeline {
    this.outputHandler = handler;
    return this;
  }

  bindToWeb(response: WebResponse, streamHandler?: WebStreamHandler): Pipeline {
    this.webAdapter = new WebOutputAdapter(response, streamHandler);
    return this;
  }

  async processWebRequest(request: WebRequest): Promise<void> {
    if (!this.webAdapter) {
      throw new Error('Pipeline not bound to web response');
    }

    try {
      const result = await this.process(
        typeof request.body === 'string' ? request.body : JSON.stringify(request.body)
      );

      // Final output
      await this.webAdapter.handleOutput(result);
      await this.webAdapter.complete();
    } catch (err) {
      const error = err instanceof Error ? err : new Error(String(err));
      await this.webAdapter.handleError(error);
    }
  }

  async process(content: string, initialMetadata: Record<string, any> = {}): Promise<ContentData> {
    let data: ContentData = {
      content,
      metadata: { ...initialMetadata }
    };

    for (const plugin of this.plugins) {
      try {
        data = await plugin.process(data, this.outputHandler);
      } catch (err) {
        const error = err instanceof Error ? err : new Error(String(err));
        if (this.outputHandler) {
          await this.outputHandler({
            type: 'error',
            content: `Error in plugin ${plugin.name}: ${error.message}`
          });
        }
        throw error;
      }
    }

    return data;
  }
}

File: core-python/src/article_dryer/pipeline.py
import json
import yaml
import logging
import asyncio
import traceback
from typing import List, Optional, Dict, Any, Union
from .types import Plugin, ContentData, OutputHandler, PipelineConfig
from .plugin_registry import PluginRegistry
from .lib.web_adapter import WebRequest, WebResponse, WebOutputAdapter, WebStreamHandler

# Setup logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class Pipeline:
    def __init__(self, plugins: List[Plugin] = None, output_handler: Optional[OutputHandler] = None):
        self.plugins = plugins or []
        self.output_handler = output_handler or self.default_output_handler
        self.web_adapter: Optional[WebOutputAdapter] = None

    async def default_output_handler(self, output):
        if self.web_adapter:
            await self.web_adapter.handle_output(output)
        elif output.type == 'error':
            print(f"ERROR: {output.content}", flush=True)
        else:
            print(json.dumps(output.content, indent=2), flush=True)

    @classmethod
    def from_config_file(cls, config_path: str, output_handler: Optional[OutputHandler] = None) -> 'Pipeline':
        with open(config_path, 'r') as f:
            config = yaml.safe_load(f)
        return cls.from_config_object(config, output_handler)

    @classmethod
    def from_config_object(cls, config: PipelineConfig, output_handler: Optional[OutputHandler] = None) -> 'Pipeline':
        registry = PluginRegistry.get_instance()
        plugins = []
        
        for plugin_config in config['plugins']:
            merged_config = {
                **config.get('globalOptions', {}),
                **plugin_config.get('options', {})
            }
            plugin = registry.create(plugin_config['name'], merged_config)
            if hasattr(plugin, 'configure'):
                plugin.configure(plugin_config.get('options', {}))
            plugins.append(plugin)
        
        return cls(plugins, output_handler)

    def add_plugin(self, plugin: Plugin) -> 'Pipeline':
        self.plugins.append(plugin)
        return self

    def set_output_handler(self, handler: OutputHandler) -> 'Pipeline':
        self.output_handler = handler
        return self

    def bind_to_web(self, response: WebResponse, stream_handler: Optional[WebStreamHandler] = None) -> 'Pipeline':
        self.web_adapter = WebOutputAdapter(response, stream_handler)
        return self

    async def process_web_request(self, request: WebRequest) -> None:
        if not self.web_adapter:
            raise RuntimeError('Pipeline not bound to web response')

        try:
            content = request.body if isinstance(request.body, str) else json.dumps(request.body)
            result = await self.process(content)
            await self.web_adapter.handle_output(result)
            await self.web_adapter.complete()
        except Exception as err:
            await self.web_adapter.handle_error(err)

    async def process(self, content: str, initial_metadata: Dict[str, Any] = None) -> ContentData:
        data = ContentData(
            content=content,
            metadata=initial_metadata or {}
        )

        for plugin in self.plugins:
            try:
                data = await plugin.process(data, self.output_handler)
            except Exception as err:
                # Print detailed stack trace
                traceback.print_exc()
                logger.error(f"Error processing plugin {plugin.name}: {str(err)}", exc_info=True)
                
                if self.output_handler:
                    await self.output_handler({
                        'type': 'error',
                        'content': f"Error in plugin {plugin.name}: {str(err)}"
                    })
                raise

        return data
